File: solutions/2833-count-zero-request-servers/solution.py
from typing import List
from collections import defaultdict

class Solution:
    def countServers(self, n: int, logs: List[List[int]], x: int, queries: List[int]) -> List[int]:
        # Sort logs by time
        logs.sort(key=lambda log: log[1])
        indexed_queries = sorted([(q, i) for i, q in enumerate(queries)])
        m = len(logs)
        q = len(queries)
        ans = [0] * q
        left = 0
        right = 0
        server_count = defaultdict(int)  # server_id -> count in window
        active_servers = set()  # servers with at least one request in window
        
        for query, idx in indexed_queries:
            # Expand right pointer to include logs with time <= query
            while right < m and logs[right][1] <= query:
                server = logs[right][0]
                server_count[server] += 1
                if server_count[server] == 1:
                    active_servers.add(server)
                right += 1
            # Move left pointer to exclude logs with time < query - x
            while left < m and logs[left][1] < query - x:
                server = logs[left][0]
                server_count[server] -= 1
                if server_count[server] == 0:
                    active_servers.discard(server)
                left += 1
            ans[idx] = n - len(active_servers)
        return ans

File: solutions/2833-count-zero-request-servers/solution.java
import java.util.*;

class Solution {
    public int[] countServers(int n, int[][] logs, int x, int[] queries) {
        // Sort logs by time
        Arrays.sort(logs, Comparator.comparingInt(log -> log[1]));
        
        // Create and sort query indices based on query values
        Integer[] queryIndices = IntStream.range(0, queries.length)
            .boxed()
            .sorted(Comparator.comparingInt(i -> queries[i]))
            .toArray(Integer[]::new);
        
        int[] result = new int[queries.length];
        int[] activeRequests = new int[n + 1]; // Assuming server IDs are 1-based
        Deque<int[]> window = new ArrayDeque<>();
        int logIndex = 0;
        int availableServers = n;
        
        for (int idx : queryIndices) {
            int leftBound = queries[idx] - x;
            int rightBound = queries[idx];
            
            // Remove logs outside the left boundary
            while (!window.isEmpty() && window.peekFirst()[1] < leftBound) {
                int serverId = window.pollFirst()[0];
                activeRequests[serverId]--;
                if (activeRequests[serverId] == 0) {
                    availableServers++;
                }
            }
            
            // Add new logs within the right boundary
            while (logIndex < logs.length && logs[logIndex][1] <= rightBound) {
                int serverId = logs[logIndex][0];
                int time = logs[logIndex][1];
                
                if (time >= leftBound) {
                    window.addLast(logs[logIndex]);
                    if (activeRequests[serverId] == 0) {
                        availableServers--;
                    }
                    activeRequests[serverId]++;
                }
                logIndex++;
            }
            
            result[idx] = availableServers;
        }
        
        return result;
    }
    

    public static void main(String[] args) {
        Solution sol = new Solution();

        int n = 3;
        int[][] logs = {{1, 1}, {2, 3}, {1, 4}, {3, 5}};
        int x = 2;
        int[] queries = {2, 3, 4, 5};
        
        int[] result = sol.countServers(n, logs, x, queries);
        System.out.println("Results: " + Arrays.toString(result));
    }
}
